library/tinkerpop/src/test/java/uk/gov/gchq/gaffer/tinkerpop/service/GafferPopNamedOperationServiceIT.java

```diff
@@ -0,0 +1,186 @@
+
+public class GafferPopNamedOperationServiceIT {
+    private static final MapStoreProperties PROPERTIES = MapStoreProperties.loadStoreProperties(StreamUtil.openStream(GafferPopGraphTest.class, "/gaffer/map-store.properties"));
+    private static final Configuration TEST_CONFIGURATION = new BaseConfiguration() {
+        {
+            this.setProperty(GafferPopGraph.GRAPH, GafferPopGraph.class.getName());
+            this.setProperty(GafferPopGraph.OP_OPTIONS, new String[] {"key1:value1", "key2:value2" });
+            this.setProperty(GafferPopGraph.USER_ID, USER_ID);
+            this.setProperty(GafferPopGraph.DATA_AUTHS, new String[]{AUTH_1, AUTH_2});
+        }
+    };
+
+    @Test
+    void shouldHaveNamedOperationService() {
+        // Given
+        final GafferPopGraph gafferPopGraph = GafferPopGraph.open(TEST_CONFIGURATION, getGafferGraph());
+        GraphTraversalSource g = gafferPopGraph.traversal();
+
+        // When
+        GraphTraversal<Object, Object> traversal = g.call().with("verbose").with("service", "namedoperation");
+        List<Object> results = traversal.toList();
+
+        // Then
+        assertThat(results).hasSize(1).asString()
+                .contains("{\"name\":\"namedoperation\",\"type:[requirements]:\":{\"Start\":[]},\"params\":{}}");
+    }
+
+    @Test
+    void shouldThrowExceptionForMissingParameter() {
+        // Given
+        final GafferPopGraph gafferPopGraph = GafferPopGraph.open(TEST_CONFIGURATION, getGafferGraph());
+        GraphTraversalSource g = gafferPopGraph.traversal();
+
+        // When
+        Map<String, String> params = Collections.singletonMap("missing", "missing");
+
+        // Then
+        assertThatThrownBy(() -> g.call("namedoperation", params).toList())
+                .isExactlyInstanceOf(IllegalStateException.class)
+                .hasMessage("Missing parameter");
+    }
+
+    @Test
+    void shouldThrowExceptionWhenTryingToExecuteMissingNamedOperation() {
+        // Given
+        GafferPopGraph gafferPopGraph = GafferPopGraph.open(TEST_CONFIGURATION, getGafferGraph());
+        GraphTraversalSource g = gafferPopGraph.traversal();
+
+        // When
+        Map<String, String> params = Collections.singletonMap("execute", "missingNamedOp");
+
+        // Then
+        assertThatThrownBy(() -> g.call("namedoperation", params).toList())
+                .isExactlyInstanceOf(IllegalStateException.class)
+                .hasMessage("Named Operation not found");
+    }
+
+    @Test
+    void shouldAddNamedOperation() {
+        // Given
+        final GafferPopGraph gafferPopGraph = GafferPopGraph.open(TEST_CONFIGURATION, getGafferGraph());
+        GraphTraversalSource g = gafferPopGraph.traversal();
+
+        // When
+        Map<String, Object> addParams = new HashMap<>();
+        addParams.put("name", "testNamedOp");
+        addParams.put("opChain", getOpChainElementCount(null).getOperationChainAsString());
```

Comment:
This adds a named op with name `null`, which I didn't know would work

Reply:
Not quite. It creates a named op which has a null name but this name is never used. The operation is created and `getOperationChainAsString()` is then used to extract only the chain from it as JSON. This chain is then used with the name "testNamedOp" to create and add a new named op.

The creation of the named operation object is only a helper to easily get the operation chain as a string, instead of having a difficult to read hardcoded JSON sting.